### PR TITLE
Jax -> Pytorch

### PR DESCRIPTION
**Pull request by ejnnr**
Very early stage draft, a lot of stuff hasn't been ported yet and straight up doesn't work. Things that should more or less work:
* Training and evaluating classifiers
* Backdoor data
* Mahalanobis training and eval
But even for these there are caveats, e.g. GPT-4 did a lot of the pytorch version of WaNet and I haven't tested that yet, and I also haven't merged in the more recent changes to validation data and the fix in #13 

I'm planning to finish porting stuff, which mainly affects abstractions and adversarial examples, clean up things that are unneeded now, and test things more extensively. I won't have much time Tue/Wed, but hopefully I'll make decent progress on those things in the second half of this week. In the meantime, hopefully this gives you a sense of what things will change @VRehnberg to avoid doing unnecessary work. (I think your recent changes will hopefully be pretty feasible to merge)

Apart from pretty direct translation from jax/flax to pytorch, the main changes are:
* Using pytorch lightning instead of the custom `Trainer` (this should give us many more advanced features that should become useful when we train bigger models)
* Replacing the way models worked before with [`HookedModel`](https://github.com/ejnnr/cupbearer/blob/a5536c470adda8eb492bcf708100b071f24bee20/src/cupbearer/models/hooked_model.py). I think this is going to make it much easier to add new models without reading a bunch of documentation, though we could try to make it even simpler.
* The activations extracted during a forward pass can now form an arbitrary acyclic graph instead of a linear one. (In terms of datastructures, I'm using dicts of activations instead of lists of activations). With the new `HookedModel` setup, there's basically no reason not to allow this more general version (it's barely more complicated to implement, and we can easily get the linear graph version as a special case).
* For now I'm removing the code to plot neural networks. I do like those plots to e.g. show where anomalies are detected, but it's much harder to do with the new `HookedModel` setup and not a high enough priority to port it right now.

### CODE REVIEW

**Comment by VRehnberg** [comment]
On Wanet the original implementation uses grid_sample instead of interpolate. Difference is probably about as big we had before so I believe that is fine. Here is my investigation from before when comparing them:
https://stackoverflow.com/questions/76961133/differences-between-cubic-2d-interpolation-methods-in-scipy-and-pytorch

**Comment by VRehnberg** [comment]
Looks promising, let me know if you want any help or when you want a proper review.

**Comment by ejnnr** [comment]
The mahalanobis detector wasn't actually working yet, I just didn't notice because one of the tests wasn't running. Fixed now, and abstractions should also work. I think the main missing pieces are:
- [ ] Porting adversarial examples (maybe now that we're using pytorch, we should use an existing library for that)
- [ ] Porting the robust optimization version of abstractions (or maybe just rip it out for now until we actually want to use it again)
- [ ] Merge in the improvements from `main`
- [ ] Ideally add some more automated tests
- [ ] Check WaNet implementation
- [ ] Check that detectors still perform similarly
- [ ] Clean-up (e.g. remove `Trainer` class)

**Comment by VRehnberg** [comment]
> Porting adversarial examples (maybe now that we're using pytorch, we should use an existing library for that)

I've seen two existing implementation while looking for defense methods [torchattack](https://github.com/Harry24k/adversarial-attacks-pytorch) and [DeepRobust](https://github.com/Harry24k/adversarial-attacks-pytorch). Haven't looked in any detail, but torchattack seems a bit better maintained.

**Comment by ejnnr** [comment]
I merged things now, but the backdoor tests are failing, I'll probably fix that next week.

For WaNet, I'm currently using `upsample` and `grid_sample` like in the original implementation, which also lead to other changes (like warping fields having values in [-1, 1] instead of [0, size]).

**Comment by VRehnberg** [comment]
> I merged things now, but the backdoor tests are failing, I'll probably fix that next week.
> 
> For WaNet, I'm currently using `upsample` and `grid_sample` like in the original implementation, which also lead to other changes (like warping fields having values in [-1, 1] instead of [0, size]).

Great. I remember scaling everything correctly was a tricky part. But, in this case it's probably easier just to try to replicate the original implementation. There's an extra factor 0.5 in the control_grid that should be removed as well now if I remember correctly, but I'll make sure to go over this part in extra detail when reviewing.

**Comment by ejnnr** [comment]
I think this is good enough to review now. My main concern is whether the methods actually work correctly (abstractions seem to perform a bit worse than in the Jax implementation). But they do at least seem to work at all, and there's a priori no strong reason to think the jax version was *more* correct than this one.

**Comment by ejnnr** [comment]
Main design uncertainties:
- How should we extract activations from models? The main alternative to the current `HookedModel` approach would be just using pytorch forward hooks on every submodule in a network. Gives somewhat less control to people implementing new models, but less overhead.
- Relatedly, how should we configure which activations to actually use? (the `names` argument in various places) I think right now this basically isn't configurable at all and just uses everything, which is pretty silly (except for abstraction, where there's a reasonable default). We could make each model automatically provide defaults. For configuration, we should probably let users specify the path to a function that will compute the names to use (statically encoding these is cumbersome if model sizes can change)

File: src/cupbearer/data/backdoors.py [hunk 1]
```diff
@@ -150,95 +144,75 @@ def warping_field(self) -> np.ndarray:
...
-                )
-                for grid in self.control_grid
-            ],
-            axis=0,
-        )
+        control_grid = self.control_grid
+        assert control_grid.ndim == 3
+        # upsample expects a batch dimesion, so we add a singleton. We permute after
+        # upsampling, since grid_sample expects the length-2 axis to be the last one.
+        field = F.upsample(
```
Comment (VRehnberg) [inline src/cupbearer/data/backdoors.py:151]:
```suggestion
        field = F.interpolate(
```
upsample has become deprecated, but should just be a straight replacement with interpolate

File: src/cupbearer/scripts/eval_classifier.py [hunk 2]
```diff
@@ -22,37 +19,20 @@ def main(cfg: Config):
...
-        optimizer=None,
-        example_input=example_input,
-        log_dir=None,
-        loggers=[],
-        enable_progress_bar=cfg.pbar,
-        rng=jax.random.PRNGKey(cfg.seed),
-        override_variables=override_variables,
+    assert cfg.dir.path is not None
+    classifier = Classifier.load_from_checkpoint(
+        cfg.dir.path / "model.ckpt", test_loader_names=["test"]
```
Comment (VRehnberg) [inline src/cupbearer/scripts/eval_classifier.py:26]:
```suggestion
        cfg.dir.path / "last.ckpt", test_loader_names=["test"]
```
This should be last, no?

**Review by VRehnberg — CHANGES_REQUESTED** [review]
Looking good, couldn't see any problem in Wanet implementation. Only found some few things overall that had yet to be updated.

Edit: Oh, and there are a gazillion warnings from lightning but, I'm not sure there's anything we can do about most of them.

**Comment by VRehnberg** [comment]
> Main design uncertainties:
> 
>     * How should we extract activations from models? The main alternative to the current `HookedModel` approach would be just using pytorch forward hooks on every submodule in a network. Gives somewhat less control to people implementing new models, but less overhead.
> 
>     * Relatedly, how should we configure which activations to actually use? (the `names` argument in various places) I think right now this basically isn't configurable at all and just uses everything, which is pretty silly (except for abstraction, where there's a reasonable default). We could make each model automatically provide defaults. For configuration, we should probably let users specify the path to a function that will compute the names to use (statically encoding these is cumbersome if model sizes can change)

It's ok for now so I think this can go into a separate PR.

 * How to extract activations?
     * Flexibility is good in the sense that ideally you should be able to use almost any subset of layers to capture activations from.
         * Though if this is done you have to be careful of memory requirements for large models. We only want to store activations from layers we're actually using right now.
         * Also, if it is this general then you might want a helper for picking these activations and storing that as a config.
     * There are a few detectors where we might want to do interventions on the activations.
          * This isn't possible with the current implementation?
     * I think forward hooks might be the way to go in the future. However,
         * there will be some work needed before it is as useful as HookedModels are now.
 * How to choose activations to use?
     * Reasonable defaults for at least implemented models
         * If we add support for using any model, then I think it is better to make choice of activations explicit for generic models.
     * A way to choose which activations to use. Pointing to a function is general enough.

File: pyproject.toml [hunk 3]
```diff
@@ -14,7 +14,6 @@ classifiers = [
     "Operating System :: OS Independent",
 ]
 
-requires-python = ">=3.10"
 dependencies = [
     "flax~=0.7.0",
```
[the remark this answers is not in the document]
Comment (ejnnr) [reply pyproject.toml]: Yep, thanks!

**Comment by ejnnr** [comment]
Notes to self:

- [ ] Fix or hide spurious warnings (at least the ones I get on my machine)
- [ ] Force models to have a method that generates default names for activations
- [ ] (optional, could also push this to another PR) Make names configurable by passing a function to scripts (or a string that's a path to a function for CLI compatibility)

After that should be fine to merge.

We'll probably switch to pytorch forward hooks, but in a separate PR.

**Comment by VRehnberg** [comment]
> Notes to self:
> 
>     * [ ]  Fix or hide spurious warnings (at least the ones I get on my machine)

Here are a (filtered) list of the type of warnings I see.

```
  /home/vikren/miniconda3/envs/cupbearer-tmp/lib/python3.10/site-packages/lightning_utilities/core/imports.py:13: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html

  /home/vikren/miniconda3/envs/cupbearer-tmp/lib/python3.10/site-packages/lightning/fabric/__init__.py:40: Deprecated call to `pkg_resources.declare_namespace('lightning.fabric')`.
  Implementing implicit namespace packages (as specified in PEP 420) is preferred to `pkg_resources.declare_namespace`. See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-packages

  /home/vikren/miniconda3/envs/cupbearer-tmp/lib/python3.10/site-packages/torchmetrics/utilities/imports.py:24: DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.
    _PYTHON_LOWER_3_8 = LooseVersion(_PYTHON_VERSION) < LooseVersion("3.8")

tests/test_data.py::test_backdoor_relabeling[WanetBackdoor]
  /home/vikren/miniconda3/envs/cupbearer-tmp/lib/python3.10/site-packages/torch/functional.py:504: UserWarning: torch.meshgrid: in an upcoming release, it will be required to pass the indexing argument. (Triggered internally at ../aten/src/ATen/native/TensorShape.cpp:3483.)
    return _VF.meshgrid(tensors, **kwargs)  # type: ignore[attr-defined]

tests/test_pipeline.py::test_eval_classifier
  /home/vikren/miniconda3/envs/cupbearer-tmp/lib/python3.10/site-packages/torch/utils/tensorboard/__init__.py:4: DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.
    if not hasattr(tensorboard, "__version__") or LooseVersion(

tests/test_pipeline.py::test_eval_classifier
  /home/vikren/miniconda3/envs/cupbearer-tmp/lib/python3.10/site-packages/lightning/pytorch/callbacks/model_checkpoint.py:634: Checkpoint directory /tmp/pytest-of-vikren/pytest-18/tests_test_pipeline_py0 exists and is not empty.

tests/test_pipeline.py::test_eval_classifier
  /home/vikren/miniconda3/envs/cupbearer-tmp/lib/python3.10/site-packages/lightning/pytorch/utilities/data.py:104: Total length of `list` across ranks is zero. Please make sure this was your intention.

  /home/vikren/miniconda3/envs/cupbearer-tmp/lib/python3.10/site-packages/lightning/pytorch/trainer/connectors/data_connector.py:441: The 'train_dataloader' does not have many workers which may be a bottleneck. Consider increasing the value of the `num_workers` argument` to `num_workers=7` in the `DataLoader` to improve performance.

  /home/vikren/miniconda3/envs/cupbearer-tmp/lib/python3.10/site-packages/lightning/pytorch/loops/fit_loop.py:293: The number of training batches (1) is smaller than the logging interval Trainer(log_every_n_steps=50). Set a lower value for log_every_n_steps if you want to see logs for the training epoch.

tests/test_pipeline.py::test_finetuning_detector
  /home/vikren/miniconda3/envs/cupbearer-tmp/lib/python3.10/site-packages/lightning/pytorch/trainer/configuration_validator.py:74: You defined a `validation_step` but have no `val_dataloader`. Skipping val loop.
```

**Review by VRehnberg — APPROVED** [review]

**Comment by ejnnr** [comment]
Haven't fixed *all* the warnings yet and several new open TODOs, but want to merge this for now since we're already building on this and this PR is too big as it is.

Note on abstractions: the reason they don't work quite as well might be that I deliberately made the abstract steps completely linear, removing the ReLU, because I'd been meaning to try that for a while (just forgot about that change). When we investigate abstractions more seriously, we should benchmark this in realistic cases.

**Pull request merged**